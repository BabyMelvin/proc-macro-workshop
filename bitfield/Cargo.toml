[package]
name = "bitfield"
version = "0.0.0"
edition = "2021"
autotests = false
publish = false

[[test]]
name = "tests"
path = "tests/progress.rs"

[[test]]
name = "bits_pos"
path = "./tests/bits_pos.rs"

[[test]]
name = "09"
path = "./tests/09-variant-out-of-range2.rs"

[dev-dependencies]
trybuild = { version = "1.0.49", features = ["diff"] }

[dependencies]
bitfield-impl = { path = "impl" }
