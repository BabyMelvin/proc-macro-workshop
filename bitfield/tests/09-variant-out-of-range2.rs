// Bitfield enums with any discriminant (implicit or explicit) outside of the
// range 0..2^BITS should fail to compile.

use bitfield::*;

const F: isize = 1;

#[derive(BitfieldSpecifier)]
enum DeliveryMode {
    Fixed = F,
    Lowest,
    SMI,
    RemoteRead,
    NMI,
    Init,
    Startup,
    External,
}

#[test]
#[should_panic]
fn enum_check_bits() {
    // 运行时检查枚举体的 discriminant 是否越界（原 09 测试是编译时检查的，而且可定位到具体的成员上）
    // dtolnay 对 bitfield 的具体实现没有过多介绍，有些 failed 类型的测试很难通过
    DeliveryMode::__check_bits(); 
}
