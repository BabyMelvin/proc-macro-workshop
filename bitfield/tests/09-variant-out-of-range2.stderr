error: any use of this value will cause an error
  --> tests/09-variant-out-of-range2.rs:18:5
   |
18 |     External,
   |     ^^^^^^^^ attempt to compute `0_u8 - 1_u8`, which would overflow
   |
   = note: `#[deny(const_err)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
